In [2]:
import pandas as pd

aggregate functions = Reduces a set of values into a single summary value.
                      Used to summarize and analyze data.
                      Often used with the groupby() function

In [3]:
#Basically we are reducing a bunch of values into a single one
df = pd.read_csv("pokemon.csv")
print(df)

       #                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generatio

We have a dataframe that contains lists of all pokemon from Gen 1 kanto to Gen 6 kalos.

Now lets say we want mean of all values from a column

    Default syntax: print(df.mean())

    But this sends error because in our dataframe their contains columns that have string and we can't calculate mean of string.

So we use print(df.mean(numeric_only = True))

    We pass in the kyword argument in df.mean() called 'numeric_only'.
    We are telling pandas to find the mean of those columns that are numeric.


In [4]:
print(df.mean(numeric_only = True))

#             362.81375
Total         435.10250
HP             69.25875
Attack         79.00125
Defense        73.84250
Sp. Atk        72.82000
Sp. Def        71.90250
Speed          68.27750
Generation      3.32375
Legendary       0.08125
dtype: float64


Now what does these data represent?

From gen 1 to gen 2, The average total stat of all pokemons is 435.10.
This means most are way below the standard strong stats of 600 for pseudo legendaries and 680 for box art legendaries.

Most also have really bad atk stats with average only being 79.001. Keep in mind pokemon like Mega Rayqueza have atk stats of 150 {Which is super high}

Similar is the case for sp.atk stats as pokemon like mega mewtwo Y have sp atk stat of 194 which is highest for all legendaries.

In [5]:
#                   Another aggregate function is the sum()
print(df.sum(numeric_only=True))
#                   This sums all the numeric values of each pokemon in a column

#             290251
Total         348082
HP             55407
Attack         63201
Defense        59074
Sp. Atk        58256
Sp. Def        57522
Speed          54622
Generation      2659
Legendary         65
dtype: int64


Most of the sum data from above is useless Except one which is:

        Their are 65 legendary from pokemon generation 1 to generation 6.

In [6]:
print(df.min(numeric_only=True))            #This returns the smallest value from each  column.

#                 1
Total           180
HP                1
Attack            5
Defense           5
Sp. Atk          10
Sp. Def          20
Speed             5
Generation        1
Legendary     False
dtype: object


In [7]:
print(df.max(numeric_only = True))            #This returns the highest value from each column.

#              721
Total          780
HP             255
Attack         190
Defense        230
Sp. Atk        194
Sp. Def        230
Speed          180
Generation       6
Legendary     True
dtype: object


In [8]:
print(df.count())                               #This counts the number of rows from each column.
#                                                 count won't include any null value

#             800
Name          800
Type 1        800
Type 2        414
Total         800
HP            800
Attack        800
Defense       800
Sp. Atk       800
Sp. Def       800
Speed         800
Generation    800
Legendary     800
dtype: int64


The above aggregate functions apply to the whole data frame. Below is the way aggregate functions work to single column.

In [9]:
print(df["Attack"].mean())                  # We Simply use a subscript operator '[]' on dataframe df to specify index.

79.00125


In [10]:
print(df["HP"].max())                       # Maximun HP in the pokedex

255


In [11]:
print(df["HP"].min())                       # Minimum HP in the pokedex

1


In [12]:
print(df["Name"].count())                   # All pokemon in the dataframe.
#                                              This tell us their are 800 distinct pokemon including 
#                                                mega evolutions the df.

800


Now we can also group together different rows (that have something in common) in the dataframes. Like we can group together pokemon by their types. {.groupby()}

In [13]:
# Lets put grass type in one group, fire type in another and soon.
#       For this we use the groupby() function.

group = df.groupby("Type 1")
print(group)            #This returns a series for us to work with and display.
#                         Remember group is an object.

In [14]:
print(group["Attack"].mean())       #We called a aggregate function mean() in our group. We also selected 
#                                     the columne "Attack" for mean value. This returns mean  of attack stat of 
#                                       all the pokemon that have same type 1.

# Judging by the data Dragon type has the highest atk average. with ground at 2nd.

Type 1
Bug          70.971014
Dark         88.387097
Dragon      112.125000
Electric     69.090909
Fairy        61.529412
Fighting     96.777778
Fire         84.769231
Flying       78.750000
Ghost        73.781250
Grass        73.214286
Ground       95.750000
Ice          72.750000
Normal       73.469388
Poison       74.678571
Psychic      71.456140
Rock         92.863636
Steel        92.703704
Water        74.151786
Name: Attack, dtype: float64


In [16]:
print(group["Name"].count())            #For listing no. of pokemon based on type 1, we can also use
#                                         other columne name such as attack, Sp. atk or Sp. Def

Type 1
Bug          69
Dark         31
Dragon       32
Electric     44
Fairy        17
Fighting     27
Fire         52
Flying        4
Ghost        32
Grass        70
Ground       32
Ice          24
Normal       98
Poison       28
Psychic      57
Rock         44
Steel        27
Water       112
Name: Name, dtype: int64


we can also use .sum(), .min(), .max() values as well for the group.

So these are aggregate functions in pandas. They are used to summarize data.